// html-webpack-plugin/lib/chunksorter.js
define('html-webpack-plugin/lib/chunksorter', function(require, exports, module) {

  'use strict';
  
  var toposort = require('html-webpack-plugin/node_modules/toposort/index');
  
  /*
    Sorts dependencies between chunks by their "parents" attribute.
  
    This function sorts chunks based on their dependencies with each other.
    The parent relation between chunks as generated by Webpack for each chunk
    is used to define a directed (and hopefully acyclic) graph, which is then
    topologically sorted in order to retrieve the correct order in which
    chunks need to be embedded into HTML. A directed edge in this graph is
    describing a "is parent of" relationship from a chunk to another (distinct)
    chunk. Thus topological sorting orders chunks from bottom-layer chunks to
    highest level chunks that use the lower-level chunks.
  
    @param {Array} chunks an array of chunks as generated by the html-webpack-plugin.
    It is assumed that each entry contains at least the properties "id"
    (containing the chunk id) and "parents" (array containing the ids of the
    parent chunks).
  
    @return {Array} A topologically sorted version of the input chunks
  */
  module.exports.dependency = function (chunks) {
    if (!chunks) {
      return chunks;
    }
  
    // We build a map (chunk-id -> chunk) for faster access during graph building.
    var nodeMap = {};
  
    chunks.forEach(function (chunk) {
      nodeMap[chunk.id] = chunk;
    });
  
    // Next, we add an edge for each parent relationship into the graph
    var edges = [];
  
    chunks.forEach(function (chunk) {
      if (chunk.parents) {
        // Add an edge for each parent (parent -> child)
        chunk.parents.forEach(function (parentId) {
          var parentChunk = nodeMap[parentId];
          // If the parent chunk does not exist (e.g. because of an excluded chunk)
          // we ignore that parent
          if (parentChunk) {
            edges.push([parentChunk, chunk]);
          }
        });
      }
    });
    // We now perform a topological sorting on the input chunks and built edges
    return toposort.array(chunks, edges);
  };
  
  /**
   * Sorts the chunks based on the chunk id.
   *
   * @param  {Array} chunks the list of chunks to sort
   * @return {Array} The sorted list of chunks
   */
  module.exports.id = function (chunks) {
    return chunks.sort(function orderEntryLast(a, b) {
      if (a.entry !== b.entry) {
        return b.entry ? 1 : -1;
      } else {
        return b.id - a.id;
      }
    });
  };
  
  /**
   * Performs identity mapping (no-sort).
   * @param  {Array} chunks the chunks to sort
   * @return {Array} The sorted chunks
   */
  module.exports.none = function (chunks) {
    return chunks;
  };
  
  /**
   * Defines the default sorter.
   */
  module.exports.auto = module.exports.id;
  
  // In webpack 2 the ids have been flipped.
  // Therefore the id sort doesn't work the same way as it did for webpack 1
  // Luckily the dependency sort is working as expected
  if ({
    "name": "webpack",
    "version": "1.13.3",
    "author": {
      "name": "Tobias Koppers @sokra"
    },
    "description": "Packs CommonJs/AMD modules for the browser. Allows to split your codebase into multiple bundles, which can be loaded on demand. Support loaders to preprocess files, i.e. json, jade, coffee, css, less, ... and your custom stuff.",
    "dependencies": {
      "async": "^1.3.0",
      "clone": "^1.0.2",
      "enhanced-resolve": "~0.9.0",
      "acorn": "^3.0.0",
      "interpret": "^0.6.4",
      "loader-utils": "^0.2.11",
      "memory-fs": "~0.3.0",
      "mkdirp": "~0.5.0",
      "node-libs-browser": "^0.6.0",
      "optimist": "~0.6.0",
      "supports-color": "^3.1.0",
      "tapable": "~0.1.8",
      "uglify-js": "~2.7.3",
      "watchpack": "^0.2.1",
      "webpack-core": "~0.6.0"
    },
    "license": "MIT",
    "devDependencies": {
      "benchmark": "^1.0.0",
      "bundle-loader": "~0.5.0",
      "codecov.io": "^0.1.2",
      "coffee-loader": "~0.7.1",
      "coffee-script": "^1.10.0",
      "component-webpack-plugin": "~0.2.0",
      "coveralls": "^2.11.2",
      "css-loader": "~0.15.0",
      "diff": "^2.0.2",
      "eslint": "^1.1.0",
      "eslint-plugin-nodeca": "^1.0.3",
      "express": "~4.13.1",
      "extract-text-webpack-plugin": "~0.8.0",
      "file-loader": "~0.8.0",
      "glob": "^5.0.14",
      "i18n-webpack-plugin": "~0.2.0",
      "istanbul": "^0.3.13",
      "jade": "^1.11.0",
      "jade-loader": "~0.7.0",
      "js-beautify": "^1.5.10",
      "json-loader": "~0.5.1",
      "less": "^2.5.1",
      "less-loader": "^2.0.0",
      "mocha": "~2.2.0",
      "mocha-lcov-reporter": "0.0.2",
      "raw-loader": "~0.5.0",
      "script-loader": "~0.6.0",
      "should": "^7.0.2",
      "style-loader": "~0.12.0",
      "url-loader": "~0.5.0",
      "val-loader": "~0.5.0",
      "vm-browserify": "~0.0.0",
      "webpack-dev-middleware": "^1.0.0",
      "worker-loader": "~0.6.0"
    },
    "engines": {
      "node": ">=0.6"
    },
    "repository": {
      "type": "git",
      "url": "git+https://github.com/webpack/webpack.git"
    },
    "homepage": "https://github.com/webpack/webpack",
    "main": "lib/webpack.js",
    "web": "lib/webpack.web.js",
    "bin": {
      "webpack": "./bin/webpack.js"
    },
    "files": [
      "lib/",
      "bin/",
      "buildin/",
      "hot/",
      "web_modules/"
    ],
    "scripts": {
      "pretest": "npm run lint && npm run beautify-lint",
      "test": "mocha",
      "travis": "npm run cover -- --report lcovonly",
      "lint": "eslint lib bin hot scripts",
      "beautify-lint": "node ./scripts/beautify-check",
      "beautify": "node ./scripts/beautify-rewrite",
      "precover": "npm run lint && npm run beautify-lint",
      "cover": "istanbul cover -x **/*.runtime.js node_modules/mocha/bin/_mocha",
      "publish-patch": "npm run lint && npm run beautify-lint && mocha && npm version patch && git push && git push --tags && npm publish"
    },
    "gitHead": "1a5fd7e8caeef794f67e56d9d734998209b0838c",
    "bugs": {
      "url": "https://github.com/webpack/webpack/issues"
    },
    "_id": "webpack@1.13.3",
    "_shasum": "e79c46fe5a37c5ca70084ba0894c595cdcb42815",
    "_from": "webpack@>=1.13.3 <2.0.0",
    "_npmVersion": "3.10.8",
    "_nodeVersion": "6.9.0",
    "_npmUser": {
      "name": "jhnns",
      "email": "mail@johannesewald.de"
    },
    "dist": {
      "shasum": "e79c46fe5a37c5ca70084ba0894c595cdcb42815",
      "tarball": "https://registry.npmjs.org/webpack/-/webpack-1.13.3.tgz"
    },
    "maintainers": [
      {
        "name": "jhnns",
        "email": "mail@johannesewald.de"
      },
      {
        "name": "sokra",
        "email": "tobias.koppers@googlemail.com"
      },
      {
        "name": "thelarkinn",
        "email": "sean.larkin@cuw.edu"
      }
    ],
    "_npmOperationalInternal": {
      "host": "packages-18-east.internal.npmjs.com",
      "tmp": "tmp/webpack-1.13.3.tgz_1477491822577_0.763972872402519"
    },
    "directories": {},
    "_resolved": "https://registry.npmjs.org/webpack/-/webpack-1.13.3.tgz",
    "readme": "ERROR: No README data found!"
  }
  .version.split('.')[0] === '2') {
    module.exports.auto = module.exports.dependency;
  }

});
