// renderkid/lib/renderKid/styles/rule/DeclarationBlock.js
define('html-webpack-plugin/node_modules/pretty-error/node_modules/renderkid/lib/renderKid/styles/rule/DeclarationBlock', function(require, exports, module) {

  // Generated by CoffeeScript 1.9.3
  'use strict';
  
  var Arbitrary, DeclarationBlock, declarationClasses;
  
  module.exports = DeclarationBlock = (function () {
    var self;
  
    self = DeclarationBlock;
  
    function DeclarationBlock() {
      this._declarations = {};
    }
  
    DeclarationBlock.prototype.set = function (prop, value) {
      var key, val;
      if (typeof prop === 'object') {
        for (key in prop) {
          val = prop[key];
          this.set(key, val);
        }
        return this;
      }
      prop = self.sanitizeProp(prop);
      this._getDeclarationClass(prop).setOnto(this._declarations, prop, value);
      return this;
    };
  
    DeclarationBlock.prototype._getDeclarationClass = function (prop) {
      var cls;
      if (prop[0] === '_') {
        return Arbitrary;
      }
      if (!(cls = declarationClasses[prop])) {
        throw Error("Unkown property `" + prop + "`. Write it as `_" + prop + "` if you're defining a custom property");
      }
      return cls;
    };
  
    DeclarationBlock.sanitizeProp = function (prop) {
      return String(prop).trim();
    };
  
    return DeclarationBlock;
  })();
  
  Arbitrary = require('html-webpack-plugin/node_modules/pretty-error/node_modules/renderkid/lib/renderKid/styles/rule/declarationBlock/Arbitrary');
  
  declarationClasses = {
    color: require('html-webpack-plugin/node_modules/pretty-error/node_modules/renderkid/lib/renderKid/styles/rule/declarationBlock/Color'),
    background: require('html-webpack-plugin/node_modules/pretty-error/node_modules/renderkid/lib/renderKid/styles/rule/declarationBlock/Background'),
    width: require('html-webpack-plugin/node_modules/pretty-error/node_modules/renderkid/lib/renderKid/styles/rule/declarationBlock/Width'),
    height: require('html-webpack-plugin/node_modules/pretty-error/node_modules/renderkid/lib/renderKid/styles/rule/declarationBlock/Height'),
    bullet: require('html-webpack-plugin/node_modules/pretty-error/node_modules/renderkid/lib/renderKid/styles/rule/declarationBlock/Bullet'),
    display: require('html-webpack-plugin/node_modules/pretty-error/node_modules/renderkid/lib/renderKid/styles/rule/declarationBlock/Display'),
    margin: require('html-webpack-plugin/node_modules/pretty-error/node_modules/renderkid/lib/renderKid/styles/rule/declarationBlock/Margin'),
    marginTop: require('html-webpack-plugin/node_modules/pretty-error/node_modules/renderkid/lib/renderKid/styles/rule/declarationBlock/MarginTop'),
    marginLeft: require('html-webpack-plugin/node_modules/pretty-error/node_modules/renderkid/lib/renderKid/styles/rule/declarationBlock/MarginLeft'),
    marginRight: require('html-webpack-plugin/node_modules/pretty-error/node_modules/renderkid/lib/renderKid/styles/rule/declarationBlock/MarginRight'),
    marginBottom: require('html-webpack-plugin/node_modules/pretty-error/node_modules/renderkid/lib/renderKid/styles/rule/declarationBlock/MarginBottom'),
    padding: require('html-webpack-plugin/node_modules/pretty-error/node_modules/renderkid/lib/renderKid/styles/rule/declarationBlock/Padding'),
    paddingTop: require('html-webpack-plugin/node_modules/pretty-error/node_modules/renderkid/lib/renderKid/styles/rule/declarationBlock/PaddingTop'),
    paddingLeft: require('html-webpack-plugin/node_modules/pretty-error/node_modules/renderkid/lib/renderKid/styles/rule/declarationBlock/PaddingLeft'),
    paddingRight: require('html-webpack-plugin/node_modules/pretty-error/node_modules/renderkid/lib/renderKid/styles/rule/declarationBlock/PaddingRight'),
    paddingBottom: require('html-webpack-plugin/node_modules/pretty-error/node_modules/renderkid/lib/renderKid/styles/rule/declarationBlock/PaddingBottom')
  };

});
