// webpack-core/lib/RawSource.js
define('webpack/node_modules/webpack-core/lib/RawSource', function(require, exports, module) {

  /*
  	MIT License http://www.opensource.org/licenses/mit-license.php
  	Author Tobias Koppers @sokra
  */
  "use strict";
  
  var Source = require("webpack/node_modules/webpack-core/lib/Source");
  var SourceNode = require("fis3-hook-commonjs/node_modules/jstransform/node_modules/source-map/lib/source-map").SourceNode;
  var SourceListMap = require("webpack/node_modules/webpack-core/node_modules/source-list-map/lib/index").SourceListMap;
  
  function RawSource(value) {
  	Source.call(this);
  	this._value = value;
  }
  module.exports = RawSource;
  
  RawSource.prototype = Object.create(Source.prototype);
  RawSource.prototype.constructor = RawSource;
  
  RawSource.prototype.source = function () {
  	return this._value;
  };
  
  RawSource.prototype.map = function (options) {
  	return null;
  };
  
  RawSource.prototype.node = function (options) {
  	return new SourceNode(null, null, null, this._value);
  };
  
  RawSource.prototype.listMap = function (options) {
  	return new SourceListMap(this._value);
  };
  
  RawSource.prototype.updateHash = function (hash) {
  	hash.update(this._value);
  };

});
