// webpack-core/lib/Source.js
define('webpack/node_modules/webpack-core/lib/Source', function(require, exports, module) {

  /*
  	MIT License http://www.opensource.org/licenses/mit-license.php
  	Author Tobias Koppers @sokra
  */
  "use strict";
  
  var SourceNode = require("fis3-hook-commonjs/node_modules/jstransform/node_modules/source-map/lib/source-map").SourceNode;
  var SourceMapConsumer = require("fis3-hook-commonjs/node_modules/jstransform/node_modules/source-map/lib/source-map").SourceMapConsumer;
  
  function Source() {}
  
  module.exports = Source;
  
  Source.prototype.source = null;
  
  Source.prototype.size = function () {
  	return this.source().length;
  };
  
  Source.prototype.map = function (options) {
  	return null;
  };
  
  Source.prototype.sourceAndMap = function (options) {
  	return {
  		source: this.source(),
  		map: this.map()
  	};
  };
  
  Source.prototype.node = null;
  
  Source.prototype.listNode = null;
  
  Source.prototype.updateHash = function (hash) {
  	var source = this.source();
  	hash.update(source || "");
  };

});
