// react/lib/React.js
define('react/lib/React', function(require, exports, module) {

  /**
   * Copyright 2013-present, Facebook, Inc.
   * All rights reserved.
   *
   * This source code is licensed under the BSD-style license found in the
   * LICENSE file in the root directory of this source tree. An additional grant
   * of patent rights can be found in the PATENTS file in the same directory.
   *
   */
  
  'use strict';
  
  var _assign = require('babel-loader/node_modules/object-assign/index');
  
  var ReactChildren = require('react/lib/ReactChildren');
  var ReactComponent = require('react/lib/ReactComponent');
  var ReactPureComponent = require('react/lib/ReactPureComponent');
  var ReactClass = require('react/lib/ReactClass');
  var ReactDOMFactories = require('react/lib/ReactDOMFactories');
  var ReactElement = require('react/lib/ReactElement');
  var ReactPropTypes = require('react/lib/ReactPropTypes');
  var ReactVersion = require('react/lib/ReactVersion');
  
  var onlyChild = require('react/lib/onlyChild');
  var warning = require('react/node_modules/fbjs/lib/warning');
  
  var createElement = ReactElement.createElement;
  var createFactory = ReactElement.createFactory;
  var cloneElement = ReactElement.cloneElement;
  
  if ('development' !== 'production') {
    var ReactElementValidator = require('react/lib/ReactElementValidator');
    createElement = ReactElementValidator.createElement;
    createFactory = ReactElementValidator.createFactory;
    cloneElement = ReactElementValidator.cloneElement;
  }
  
  var __spread = _assign;
  
  if ('development' !== 'production') {
    var warned = false;
    __spread = function () {
      'development' !== 'production' ? warning(warned, 'React.__spread is deprecated and should not be used. Use ' + 'Object.assign directly or another helper function with similar ' + 'semantics. You may be seeing this warning due to your compiler. ' + 'See https://fb.me/react-spread-deprecation for more details.') : void 0;
      warned = true;
      return _assign.apply(null, arguments);
    };
  }
  
  var React = {
  
    // Modern
  
    Children: {
      map: ReactChildren.map,
      forEach: ReactChildren.forEach,
      count: ReactChildren.count,
      toArray: ReactChildren.toArray,
      only: onlyChild
    },
  
    Component: ReactComponent,
    PureComponent: ReactPureComponent,
  
    createElement: createElement,
    cloneElement: cloneElement,
    isValidElement: ReactElement.isValidElement,
  
    // Classic
  
    PropTypes: ReactPropTypes,
    createClass: ReactClass.createClass,
    createFactory: createFactory,
    createMixin: function createMixin(mixin) {
      // Currently a noop. Will be used to validate and trace mixins.
      return mixin;
    },
  
    // This looks DOM specific but these are actually isomorphic helpers
    // since they are just generating DOM strings.
    DOM: ReactDOMFactories,
  
    version: ReactVersion,
  
    // Deprecated hook for JSX spread, don't use this for anything.
    __spread: __spread
  };
  
  module.exports = React;

});
